builder/api/src/main/java/io/helidon/builder/api/ProvidedUtil.java

```
@@ -0,0 +1,192 @@
+            return servicesFromList(serviceLoader, providerType, configType, configuredServices, discoverServices);
+        } else {
+            // driven by service loader order
+            return servicesFromObject(serviceLoader, providerType, configType, configuredServices, discoverServices);
+        }
+    }
+
+    private static <T extends NamedService> List<T>
+    servicesFromObject(HelidonServiceLoader<? extends ConfiguredProvider<T>> serviceLoader,
+                       Class<? extends ConfiguredProvider<T>> providerType,
+                       Class<T> configType,
+                       List<ConfiguredService> configuredServices,
+                       boolean allFromServiceLoader) {
+        // order is determined by service loader
+        Set<String> availableProviders = new HashSet<>();
+        Map<String, ConfiguredService> allConfigs = new HashMap<>();
+        configuredServices.forEach(it -> allConfigs.put(it.type(), it));
+        Set<String> unusedConfigs = new HashSet<>(allConfigs.keySet());
+
+        List<T> result = new ArrayList<>();
+
+        serviceLoader.forEach(provider -> {
+            ConfiguredService configuredService = allConfigs.get(provider.configKey());
+            availableProviders.add(provider.configKey());
+            unusedConfigs.remove(provider.configKey());
+            if (configuredService == null) {
+                if (allFromServiceLoader) {
+                    result.add(provider.create(Config.empty(), provider.configKey()));
+                }
+            } else {
+                if (configuredService.enabled()) {
+                    result.add(provider.create(configuredService.serviceConfig(),
+                                               configuredService.name()));
+                }
+            }
+        });
+        if (!unusedConfigs.isEmpty()) {
+            throw new ConfigException("Unknown provider configured. Expected providers with types: " + unusedConfigs
```

why is this always a hard exception, instead of cases where we generate a info or warning but are allowed to move on? seems to aggressive in nature to me.

---

Because you have explicitly defined a service you want to use, but it is not available. This is a configuration error, and silently ignoring it would cause trouble later on.

---

We do the same thing in security - if you define a security provider in configuration, but do not have it configured, it is an error

---

I still feel there should be a flag for this that we can toggle.

---

If you find a use case in Helidon where this is correct, please add an issue and we will do it. Right now this feels like the correct "fail-fast" principle.